In [31]:
import pandas as pd

## Loading data


In [32]:
df_historical_data = pd.read_csv('datasets/fifa_worldcup_matches.csv')
df_fixture = pd.read_csv('datasets/fifa_worldcup_fixture.csv')
df_missing_data = pd.read_csv('datasets/fifa_worldcup_missing_data.csv')



In [33]:
df_historical_data

home         score      away  year
0       France            4–1    Mexico  1930
1    Argentina            1–0    France  1930
2        Chile            3–0    Mexico  1930
3        Chile            1–0    France  1930
4    Argentina            6–3    Mexico  1930
..          ...           ...       ...   ...
452     Russia   2–2 (a.e.t.)   Croatia  2018
453     France            1–0   Belgium  2018
454    Croatia   2–1 (a.e.t.)   England  2018
455    Belgium            2–0   England  2018
456     France            4–2   Croatia  2018

[457 rows x 4 columns]

In [34]:
df_fixture

home     score              away  year
0             Qatar    Match 1           Ecuador  2022
1           Senegal    Match 2       Netherlands  2022
2             Qatar   Match 18           Senegal  2022
3       Netherlands   Match 19           Ecuador  2022
4           Ecuador   Match 35           Senegal  2022
..               ...       ...               ...   ...
59  Winners Match 51  Match 59  Winners Match 52  2022
60  Winners Match 57  Match 61  Winners Match 58  2022
61  Winners Match 59  Match 62  Winners Match 60  2022
62   Losers Match 61  Match 63   Losers Match 62  2022
63  Winners Match 61  Match 64  Winners Match 62  2022

[64 rows x 4 columns]

In [35]:
df_missing_data

home score           away  year
0          Mexico    0–0   Soviet Union  1970
1         Belgium    3–0    El Salvador  1970
2    Soviet Union    4–1        Belgium  1970
3          Mexico    4–0    El Salvador  1970
4    Soviet Union    2–0    El Salvador  1970
..             ...   ...            ...   ...
503         Spain    0–1    Switzerland  2010
504         Chile    1–0    Switzerland  2010
505         Spain    2–0       Honduras  2010
506         Chile    1–2          Spain  2010
507   Switzerland    0–0       Honduras  2010

[508 rows x 4 columns]

## Cleaning data

### Cleaning white spaces and new lines for each value in the home and away features

In [36]:
df_fixture['home'] = df_fixture['home'].str.strip()
df_fixture['away'] = df_fixture['away'].str.strip()

### Cleaning missing data df

In [37]:
df_missing_data['home'].isnull()

0      False
1      False
2      False
3      False
4      False
       ...  
503    False
504    False
505    False
506    False
507    False
Name: home, Length: 508, dtype: bool

In [38]:
df_missing_data['home']

0            Mexico 
1           Belgium 
2      Soviet Union 
3            Mexico 
4      Soviet Union 
           ...      
503           Spain 
504           Chile 
505           Spain 
506           Chile 
507     Switzerland 
Name: home, Length: 508, dtype: object

### Retrieving the rows in df that matches home = null, and eliminating NaN rows (which are 64 rows)

In [45]:
''' 
    With the following code, we can check all of the rows that contain null values
    in the Home column: 

        df_missing_data[df_missing_data['home'].isnull()] 
'''
df_missing_data.dropna(inplace=True)

### Now we have 508 - 64 = 444 rows in the df_missing_data

In [40]:
df_missing_data

home score           away  year
0          Mexico    0–0   Soviet Union  1970
1         Belgium    3–0    El Salvador  1970
2    Soviet Union    4–1        Belgium  1970
3          Mexico    4–0    El Salvador  1970
4    Soviet Union    2–0    El Salvador  1970
..             ...   ...            ...   ...
503         Spain    0–1    Switzerland  2010
504         Chile    1–0    Switzerland  2010
505         Spain    2–0       Honduras  2010
506         Chile    1–2          Spain  2010
507   Switzerland    0–0       Honduras  2010

[444 rows x 4 columns]

### Once the missing data is cleaned, we have to merge df_missing_data with df_historical_data

In [48]:
df_data_historical = pd.concat([df_historical_data, df_missing_data], ignore_index=True)
# ? ignore_index: ignores the index correlation for each df.

df_data_historical

home score          away  year
0         France    4–1        Mexico  1930
1      Argentina    1–0        France  1930
2          Chile    3–0        Mexico  1930
3          Chile    1–0        France  1930
4      Argentina    6–3        Mexico  1930
..            ...   ...           ...   ...
896        Spain    0–1   Switzerland  2010
897        Chile    1–0   Switzerland  2010
898        Spain    2–0      Honduras  2010
899        Chile    1–2         Spain  2010
900  Switzerland    0–0      Honduras  2010

[901 rows x 4 columns]

### Next, we need to eliminate duplicates

In [49]:
df_data_historical.drop_duplicates(inplace=True)
# ? inplace parameters refers that the original df will be affected.

df_data_historical.sort_values('year', inplace=True)
# ? Sorting values by year

df_data_historical

home score            away  year
0       France    4–1          Mexico  1930
17     Uruguay    4–2       Argentina  1930
16     Uruguay    6–1      Yugoslavia  1930
15   Argentina    6–1   United States  1930
14    Paraguay    1–0         Belgium  1930
..          ...   ...             ...   ...
419     Brazil    2–0      Costa Rica  2018
420     Serbia    1–2     Switzerland  2018
421     Serbia    0–2          Brazil  2018
408     France    1–0            Peru  2018
450     Brazil    1–2         Belgium  2018

[901 rows x 4 columns]

# Cleaning df_data_historical
----

Removing results that doesn´t mattch de score format (Ex. 2-0, 1-2...).

We can check the data in df_data_historical, the score value in this row is "w/o[a]". We have to remove this to maintain data consistency.

|  idx |home   |score   |away   |year   |
|---|---|---|---|---|
| 37  | Sweden  | w/o[a]  | Austria  | 1938  |



In [50]:
row_index = df_data_historical[df_data_historical['home'].str.contains('Sweden', na=False) & 
df_data_historical['away'].str.contains('Austria', na=False)].index 

# ? Stores index 37

df_data_historical.drop(index=row_index, inplace=True)

df_data_historical

home score            away  year
0       France    4–1          Mexico  1930
17     Uruguay    4–2       Argentina  1930
16     Uruguay    6–1      Yugoslavia  1930
15   Argentina    6–1   United States  1930
14    Paraguay    1–0         Belgium  1930
..          ...   ...             ...   ...
419     Brazil    2–0      Costa Rica  2018
420     Serbia    1–2     Switzerland  2018
421     Serbia    0–2          Brazil  2018
408     France    1–0            Peru  2018
450     Brazil    1–2         Belgium  2018

[900 rows x 4 columns]

### Deleting (a.e.t.)

Some of the scores have the following format: 
    [number-number] [(a.e.t.)]

Example: "1-0 (a.e.t.)"

We proceed by replacing "(a.e.t.)" with an empty space

In [51]:
df_data_historical['score'] = df_data_historical['score'].str.replace('(a.e.t)', '', regex=True)
df_data_historical['score'] = df_data_historical['score'].str.replace('(.)', '')


### Removing white spaces

In [52]:
df_data_historical['home'] = df_data_historical['home'].str.strip() 
df_data_historical['away'] = df_data_historical['away'].str.strip() 

In [53]:
df_data_historical

home score           away  year
0       France   4–1         Mexico  1930
17     Uruguay   4–2      Argentina  1930
16     Uruguay   6–1     Yugoslavia  1930
15   Argentina   6–1  United States  1930
14    Paraguay   1–0        Belgium  1930
..         ...   ...            ...   ...
419     Brazil   2–0     Costa Rica  2018
420     Serbia   1–2    Switzerland  2018
421     Serbia   0–2         Brazil  2018
408     France   1–0           Peru  2018
450     Brazil   1–2        Belgium  2018

[900 rows x 4 columns]

### Separating score

Since the format of the score is, for example, "1-2". We wanna separate de goals qty for each team in separate columns. For this, we are gonna use the **split** function.

In [54]:
df_data_historical[['home_goals', 'away_goals']] = df_data_historical['score'].str.split('–', expand=True)
# ? expand parameter is to separate in different columns
# ! Each of these columns are inserted directly as new columns in the original df.
# ? At this point, the score column can be deleted.


df_data_historical.drop('score', axis=1, inplace=True)

# ? Renaming columns
df_data_historical.rename(columns={'home': 'HomeTeam', 'away': 'AwayTeam', 'home_goals':'HomeGoals', 'away_goals': 'AwayGoals', 'year': 'Year'}, inplace=True)

df_data_historical

HomeTeam       AwayTeam  Year HomeGoals AwayGoals
0       France         Mexico  1930         4         1
17     Uruguay      Argentina  1930         4         2
16     Uruguay     Yugoslavia  1930         6         1
15   Argentina  United States  1930         6         1
14    Paraguay        Belgium  1930         1         0
..         ...            ...   ...       ...       ...
419     Brazil     Costa Rica  2018         2         0
420     Serbia    Switzerland  2018         1         2
421     Serbia         Brazil  2018         0         2
408     France           Peru  2018         1         0
450     Brazil        Belgium  2018         1         2

[900 rows x 5 columns]

## Setting types

Setting Home and Away goals as int64, since currently these have "object" as data type.

In [55]:
df_data_historical.dtypes

HomeTeam     object
AwayTeam     object
Year          int64
HomeGoals    object
AwayGoals    object
dtype: object

In [56]:
df_data_historical = df_data_historical.astype({'HomeGoals': int, 'AwayGoals': int})
df_data_historical.dtypes

HomeTeam     object
AwayTeam     object
Year          int64
HomeGoals     int64
AwayGoals     int64
dtype: object

## Practice - Finding total goals in a match

For this, we're gonna add an extra column for the sum of both score columns.

In [57]:
df_data_historical['TotalGoals'] = df_data_historical['HomeGoals'] + df_data_historical['AwayGoals']
df_data_historical

HomeTeam       AwayTeam  Year  HomeGoals  AwayGoals  TotalGoals
0       France         Mexico  1930          4          1           5
17     Uruguay      Argentina  1930          4          2           6
16     Uruguay     Yugoslavia  1930          6          1           7
15   Argentina  United States  1930          6          1           7
14    Paraguay        Belgium  1930          1          0           1
..         ...            ...   ...        ...        ...         ...
419     Brazil     Costa Rica  2018          2          0           2
420     Serbia    Switzerland  2018          1          2           3
421     Serbia         Brazil  2018          0          2           2
408     France           Peru  2018          1          0           1
450     Brazil        Belgium  2018          1          2           3

[900 rows x 6 columns]

# Exporting DF to CSV

Since we have already cleaned the data, we proceed to export the dataframe to a CSV file.

Setting argument **index=False**, ignores index column when the csv is generated.

In [60]:
import os

cleaned_datasets_path = "cleaned_datasets"

if not os.path.exists(cleaned_datasets_path):
    os.makedirs(cleaned_datasets_path)

In [61]:
df_data_historical.to_csv('cleaned_datasets/cleaned_fifa_worldcup_matches.csv', index=False)
df_fixture.to_csv('cleaned_datasets/cleaned_fifa_worldcup_fixtures.csv', index=False)